src/lib.rs

```diff
@@ -1,6 +1,210 @@
+//! means that the system call will match, regardless of the actual arguments.
+//!
+//! # Example
+//!
+//! The following example defines and installs a simple filter, that sends SIGSYS for `accept4`,
+//! `fcntl(any, F_SETFD, FD_CLOEXEC, ..)` and `fcntl(any, F_GETFD, ...)`.
+//! It allows any other syscalls.
+//!
+//! ```
+//! use std::convert::TryInto;
+//! use seccompiler::{
+//!     BpfProgram, SeccompAction, SeccompCmpArgLen, SeccompCmpOp, SeccompCondition,
+//!     SeccompFilter, SeccompRule
+//! };
+//!
+//! let filter: BpfProgram = SeccompFilter::new(
+//!     vec![
+//!         (libc::SYS_accept4, vec![]),
+//!         (
+//!             libc::SYS_fcntl,
+//!             vec![
+//!                 SeccompRule::new(vec![
+//!                     SeccompCondition::new(
+//!                         1,
+//!                       SeccompCmpArgLen::Dword,
+//!                         SeccompCmpOp::Eq,
+//!                         libc::F_SETFD as u64,
+//!                     ).unwrap(),
+//!                     SeccompCondition::new(
+//!                         2,
+//!                         SeccompCmpArgLen::Dword,
+//!                         SeccompCmpOp::Eq,
+//!                         libc::FD_CLOEXEC as u64,
+//!                     ).unwrap(),
+//!                 ]).unwrap(),
+//!                 SeccompRule::new(vec![SeccompCondition::new(
+//!                     1,
+//!                     SeccompCmpArgLen::Dword,
+//!                     SeccompCmpOp::Eq,
+//!                     libc::F_GETFD as u64,
+//!                 ).unwrap()]).unwrap(),
+//!             ],
+//!         ),
+//!     ]
+//!     .into_iter()
+//!     .collect(),
+//!     SeccompAction::Allow,
+//!     SeccompAction::Trap,
+//!     std::env::consts::ARCH.try_into().unwrap(),
+//! ).unwrap().try_into().unwrap();
+//!
+//! seccompiler::apply_filter(&filter).unwrap();
+//! ```
+//!
+//! [`SeccompFilter`]: struct.SeccompFilter.html
+//! [`SeccompCondition`]: struct.SeccompCondition.html
+//! [`SeccompRule`]: struct.SeccompRule.html
+//! [`SeccompAction`]: struct.SeccompAction.html
+//!
+
+mod backend;
+
+use std::fmt::{Display, Formatter};
+
+// Re-export the IR public types.
+pub use backend::{
+    sock_filter, BpfProgram, BpfProgramRef, Error as BackendError, SeccompAction, SeccompCmpArgLen,
+    SeccompCmpOp, SeccompCondition, SeccompFilter, SeccompRule, TargetArch,
+};
+
+// BPF structure definition for filter array.
+// See /usr/include/linux/filter.h .
+#[repr(C)]
+struct sock_fprog {
+    pub len: ::std::os::raw::c_ushort,
+    pub filter: *const sock_filter,
+}
+
+/// Library Result type.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Library errors.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// Error originating in the backend compiler.
+    Backend(BackendError),
+    /// Attempting to install an empty filter.
+    EmptyFilter,
+    /// System error related to calling `prctl`.
+    Prctl(i32),
```

Comment:
This makes it harder to read the errors. Why aren't we using `std::io`? If we need the `PartialEq` which is a good reason in itself, we should improve the error message and say that the arbitrary number is an errno.

Reply:
good idea. Replaced this with `std::io::Error`. The `PartialEq` is not a hard requirement. It was just useful when testing but it can be worked around